### PR TITLE
Direct storage of split nodes

### PR DESCRIPTION
**Pull request by ehildenb**
~Blocked on: https://github.com/runtimeverification/pyk/pull/122~
~Blocked on: https://github.com/runtimeverification/pyk/pull/294~
~Blocked on: https://github.com/runtimeverification/pyk/pull/302~
~Blocked on: https://github.com/runtimeverification/pyk/pull/303~

This stores Split nodes as their own data-structure, rather than shimming them in as edges/covers. As I've been working on the summarizer paper, it's apparent that we need more restrictions on what a valid KCFG is. Explicit representation of split edges allows us to do that.

**NOTE**: We _lose_ the ability to represent non-determinism properly here! This is because we are restricting each node to having _only one_ outgoing Edge, Cover, or Split from it! We can recover this later with direct representation of non-determinism, as a multi-edge that takes exactly one step on each edge to the non-deterministic transitions.

- Add `KCFG.Successor` class, which can be an edge or a multi-edge, and just has the `source` attribute. Make `KCFG.EdgeLike` a subclass of this class.
- Add `KCFG.Split` class, which is a subclass of `KCFG.Successor`, and stores a list of `targets`. Each `target` is a `Tuple[Node, CSubst]`, which is basically a "reverse cover", it points at the more specialized node instead of the more abstract node.
- Stricter enforcement on the structure of the `KCFG`. In particular, every node can have at _most_ one successor. We were allowing multiple 0-step `KCFG.Edge` out of a given node to represent case splits. Those are directly represented as the `KCFG.Split` multi-edge now.
- Adjusts the various routines to use the new `KCFG.split` datastructure, including:
  - `reachable_nodes`: traverses splits forward/backwards.
  - `paths_between`: almost complete rewrite now that we must support multi-edges.
  - `KCFG.pretty_segments` now directly detects splits instead of implicitly detecting them as "multiple outgoing 0-depth `KCFG.Edge` on the same node".
- Updates to `KCFGExplore` routines:
  - `KCFGExplore.step` is now allowed to assume only one successor edge, simplifying its logic.
  - `KCFGExplore.all_path_reachability_prove` can now take advantage of new `KCFG.split_on_constraints`, which stores the split as a `KCFG.Split` node.
- Adds a test of reachable nodes in the reverse case, that was failing before fixing a bug in the handling of `KCFG.Split` nodes.

### CODE REVIEW

**Comment by ehildenb** [comment]
@nishantjr and @h0nzZik I've pinged you as well so that you are aware of the new restrictions on KCFGs, notably that we are storing split edges directly, and that each node has exactly 1 successor.

**Comment by nishantjr** [comment]
Nice! I like this. Now each node has a clear semantic meaning.

File: src/pyk/kcfg/kcfg.py [hunk 1]
```diff
@@ -649,17 +688,15 @@ def replace_node(self, node_id: str, new_cterm: CTerm) -> str:
 
         return new_node.id
 
+    def successors(self, source_id: str) -> Iterable[Successor]:
+        out_edges = cast(Iterable[KCFG.Successor], self.edges(source_id=source_id))
+        out_covers = cast(Iterable[KCFG.Successor], self.covers(source_id=source_id))
+        out_splits = cast(Iterable[KCFG.Successor], self.splits(source_id=source_id))
+        return chain(out_edges, out_covers, out_splits)
```
Comment (tothtamas28) [inline src/pyk/kcfg/kcfg.py:695]: I'd just return a list, on call site the result is typically converted to a list or traversed fully anyways.

**Review by tothtamas28 — APPROVED** [review]

File: src/pyk/kcfg/kcfg.py [hunk 2]
```diff
@@ -38,40 +36,37 @@ def id(self) -> str:
...
         def pretty(self, kprint: KPrint) -> Iterable[str]:
             ...
 
-        def __lt__(self, other: Any) -> bool:
-            if not isinstance(other, KCFG.EdgeLike):
-                return NotImplemented
-            return (self.source, self.target) < (other.source, other.target)
+    class EdgeLike(Successor):
+        source: 'KCFG.Node'
+        target: 'KCFG.Node'
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kcfg/kcfg.py:53]: Issue made: https://github.com/runtimeverification/pyk/issues/306

File: src/pyk/kcfg/kcfg.py [hunk 3]
```diff
@@ -213,6 +241,7 @@ def to_dict(self) -> Dict[str, Any]:
         target = sorted(self._target)
         expanded = sorted(self._expanded)
         aliases = dict(sorted(self._aliases.items()))
+        splits = dict(sorted((k, s.to_dict()) for k, s in self._splits.items()))
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kcfg/kcfg.py:244]: Done.

File: src/pyk/kcfg/kcfg.py [hunk 4]
```diff
@@ -649,17 +688,15 @@ def replace_node(self, node_id: str, new_cterm: CTerm) -> str:
 
         return new_node.id
 
+    def successors(self, source_id: str) -> Iterable[Successor]:
+        out_edges = cast(Iterable[KCFG.Successor], self.edges(source_id=source_id))
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kcfg/kcfg.py:692]: Done.

File: src/pyk/kcfg/kcfg.py (hunk 1 above)
Comment (ehildenb) [reply src/pyk/kcfg/kcfg.py:695]: Done.

File: src/pyk/kcfg/kcfg.py [hunk 5]
```diff
@@ -108,9 +103,41 @@ def pretty(self, kprint: KPrint, minimize: bool = True) -> Iterable[str]:
...
+
+        def to_dict(self) -> Dict[str, Any]:
+            return {
+                'source': self.source.id,
+                'targets': {target.id: csubst.to_dict() for target, csubst in self.targets},
+            }
+
+        @property
+        def target_ids(self) -> List[str]:
+            return sorted([t.id for t, _ in self.targets])
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kcfg/kcfg.py]: Done.

**Pull request merged**